feat(lsp): completionItem/labelDetailsSupport

rmuir opened the pull request:
PROBLEM: nvim doesn't advertise labelDetailsSupport, which leads to overly verbose 'menu' data in the PUM. For some language servers this means PUM always fills the entire horizontal width of the screen: often leaving no room to see "kind" and usually no room at all for any documentation popup.

SOLUTION: enable labelDetailsSupport, which can allow for a more compact completion PUM by default when servers support it, and for improved customization via user convert() functions.

Background: I've been trying to help test #32820, but this is a huge usability issue for a couple of the language servers out-of-box (Examples here are in C and in Java). labelDetailsSupport is the only cleanish way I know to solve it, so that there's actually room on the screen for LSP documentation popup.

Before:
![Screen_Shot_2025-03-15_at_22 14 53](https://github.com/user-attachments/assets/7854bd46-0833-4cba-9ec1-3f9815fe9415)
![Screen_Shot_2025-03-15_at_22 40 09](https://github.com/user-attachments/assets/716bf1ac-dfa7-4965-bcd2-6d14d983130d)

After:
![Screen_Shot_2025-03-15_at_22 42 27](https://github.com/user-attachments/assets/82b6b454-9b4a-4e93-a1b3-271b53de6c2b)
![Screen_Shot_2025-03-15_at_22 41 25](https://github.com/user-attachments/assets/f4d88e33-3bfc-4fb6-8fc1-28a0710aabf4)

glepnir commented:
> labelDetailsSupport is the only cleanish way I know to solve it, so that there's actually room on the screen for LSP documentation popup.

another hack way is using `convert` field which in `completion.enable` opts  to set `menu` to empty and set label to info..but I think this pr is most good way

rmuir committed:
feat(lsp): completionItem/labelDetailsSupport

PROBLEM: nvim doesn't advertise labelDetailsSupport, which leads to
overly verbose 'abbr'/'menu' data in the PUM. For some language servers
this means PUM always fills the entire horizontal width of the screen:
often leaving no room to see "kind" and usually no room at all for any
documentation popup.

SOLUTION: enable labelDetailsSupport, which can allow for a more compact
completion PUM by default when servers support it, and for improved
customization via user convert() functions.

justinmk commented:
> For some language servers this means PUM always fills the entire horizontal width of the screen: often leaving no room to see "kind" and usually no room at all for any documentation popup.

Yeah, I was surprised that 'pumwidth' exists to set a minimum width, but not a maximum. 

Meanwhile, mitigating that with small changes like this PR, is welcome.

mfussenegger commented:
I wonder if we should couple this with some way to automatically show the signature help or make it dependent on `completeopt` containing `popup` (after the other PR is in).

The whole thing probably needs to be considered together because signature help, pum and popup all 3 compete for space.

E.g. in your example you have `setDiagnostics` with 2 different signatures, in the "After: " version the difference is no longer visible. So if you want to expand one particular version you no longer know which you have to select - making it worse if there is no other way to display that information.

rmuir commented:
signature help is a good one to think about, I struggle with space on that one myself. I trigger it:
* control-S keybind
* on server's signature help trigger characters, e.g. `(`
* on ModeChanged n-v with active snippet

I give it a huge z-index to workaround "competition".

> E.g. in your example you have `setDiagnostics` with 2 different signatures, in the "After: " version the difference is no longer visible. So if you want to expand one particular version you no longer know which you have to select - making it worse if there is no other way to display that information.

I think this is a separate problem, of which there may be many solutions? I also think it is only a problem for certain languages like java, that aren't even usable at all today. For this PR I just went with "minimal", maybe I went too minimal? With labelDetailsSupport, all that data is still there... and formatted in a more useful way. The problem is where to put it all, and I argue what we are doing today isn't great. 

Personally, i use a custom convert() function, with simple `string.format` truncation, similar to vscode. And I shove the full `details` into the popup, with horizontal separator above any documentation.  It is ugly but just showing that your issue can be solved without signature help.

![Screen_Shot_2025-03-17_at_20 11 41](https://github.com/user-attachments/assets/4e73972a-4459-40b2-949c-fca5311aab4e)

If you want, we could split this change up. I can turn on labelDetails in a way that doesn't change the pum at all, just makes the new data fields available. This just means setting `abbr` to `abbr + labelDetails.detail` so that it looks like it did before, and using full 'detail' column (which is wildly inappropriate here in the java case to show in the pum at all) instead of `labelDetails.description` which is actually documented in the spec as being something that should be displayed beside the word.

LabelDetailsSupport is useful on its own for anyone that passes their own `convert`. Then we could separately decide what to do with all the data?

mfussenegger commented:
> I also think it is only a problem for certain languages like java, that aren't even usable at all today. 

Well, it is currently usable for me :p
I think the difference is that I don't use preview or popup. So all the space is available for the completion candidates


> Personally, i use a custom convert() function, with simple string.format truncation, similar to vscode

That looks more usable to me than trimming away the full signature, but in some cases you still can't disambiguate

> If you want, we could split this change up. I can turn on labelDetails in a way that doesn't change the pum at all, just makes the new data fields available. This just means setting abbr to abbr + labelDetails.detail so that it looks like it did before

This sounds like a good way to start :+1: 
I can imagine that later on we could add the trim logic you have - maybe depending on if popup/preview is enabled.


Thinking about the signature-help a bit more: It's probably only partially related, because many servers only show it once you select a candidate on the next `(` - although there are some exceptions where a `.` can also trigger it and you'd get all three candidates + signature help + documentation popup shown.
I guess one solution for that would be to have signature-help show a minimal 1-2 line version if a pum/documentation popup is already visible. But better to keep that separate.

rmuir commented:
Yeah, let's iterate. I will change this PR to "just enable the option" and to reconstruct PUM same as we had before, when labelDetails response is returned.

I would first want to make a followup PR to that, to prefer `.labelDetails.description` for `menu` when available. I don't think this piece would be controversial, and it would improve the java case: their "full signature of implementing class" in the `.detail` isn't supposed to go in there, with labelDetails it is just the return value which is much more reasonable. It is clear win with labelDetails, as it allows the server to disambiguate exactly this.

Then we can separately figure out what to do with `abbr`...

justinmk committed:
Merge branch 'master' into completionitem_labeldetails
